adagucserverEC/CDBAdapterPostgreSQL.cpp

```diff
@@ -356,175 +357,110 @@ CDBStore::Store *CDBAdapterPostgreSQL::getFilesAndIndicesForDimensions(CDataSour
-              if (l == 0) {
-                if (!CServerParams::checkTimeFormat(sDims[l])) timeValidationError = true;
-                // Get closest lowest value to this requested one, or if request value is way below get earliest value:
-                subQuery.printconcat("%s >= (select max(%s) from %s where %s <= '%s' or %s = (select min(%s) from %s)) ", netCDFDimName.c_str(), netCDFDimName.c_str(), tableName.c_str(),
-                                     netCDFDimName.c_str(), sDims[l].c_str(), netCDFDimName.c_str(), netCDFDimName.c_str(), tableName.c_str());
-                // subQuery.printconcat("%s >= '%s' ",netCDFDimName.c_str(),sDims[l].c_str());
-              }
-              if (l == 1) {
-                if (!CServerParams::checkTimeFormat(sDims[l])) timeValidationError = true;
-                subQuery.printconcat("%s <= '%s' ", netCDFDimName.c_str(), sDims[l].c_str());
-              }
-            }
-          }
-        }
-        delete[] sDims;
-      }
-      delete[] cDims;
-    }
-    if (i == 0) {
-      if (numQueriesAdded == 0) {
-        subQuery.printconcat("where ");
-      } else {
-        subQuery.printconcat("and ");
-      }
-      if (dataSource->queryBBOX) {
+  // Filter on given dimensions
+  for (size_t i = 0; i < tableDimStore->size(); i++) {
+    const char *dimName = tableDimStore->getRecord(i)->get("dimension")->c_str();
+    const char *tableName = tableDimStore->getRecord(i)->get("tablename")->c_str();
+    CT::string *dimVals = dimMap[dimName];
 
-        subQuery.printconcat("adaguctilinglevel = %d and minx >= %f and maxx <= %f and miny >= %f and maxy <= %f ", dataSource->queryLevel, dataSource->nativeViewPortBBOX[0],
-                             dataSource->nativeViewPortBBOX[2], dataSource->nativeViewPortBBOX[1], dataSource->nativeViewPortBBOX[3]);
+    if (dimVals->count == 1) {
+      const char *dimVal = dimVals[0].c_str();
 
+      // If dimension value is a number, find closest value.
+      // FIXME: not sure if isNumeric() is a perfect subtitution for querying information_schema data_type field
+      if (dimVals[0].isNumeric()) {
+        query.printconcat("AND ABS(%s - %s) = (SELECT MIN(ABS(%s - %s)) FROM %s) ", dimVal, dimName, dimVal, dimName, tableName);
       } else {
-        subQuery.printconcat("adaguctilinglevel != %d ", -1);
+        query.printconcat("AND %s = '%s' ", dimName, dimVal);
       }
-      subQuery.printconcat("ORDER BY %s DESC limit %d)a%d ", netCDFDimName.c_str(), limit, i);
-      // subQuery.printconcat("ORDER BY %s DESC )a%d ",netCDFDimName.c_str(),i);
     } else {
-      subQuery.printconcat("ORDER BY %s DESC)a%d ", netCDFDimName.c_str(), i);
+      // Find value within range of dimVals[0] and dimVals[1]
+      // Get closest lowest value to this requested one, or if request value is way below get earliest value:
+      query.printconcat(
+        "AND %s >= (SELECT MAX(%s) FROM %s WHERE %s <= '%s' OR %s = (SELECT MIN(%s) FROM %s)) AND %s <= '%s' ",
```

Comment:
To test the functionaltests (bash runtests.sh) with postgres, you can replace globally over all xml files:
```xml
<DataBase dbtype="sqlite" parameters="{ADAGUC_TMP}/adaguc.autoresource.db"/>
```
with 
```xml
<DataBase parameters="{ADAGUC_DB}"/>
```

I noticed differences like:

```xml
<Dimension name="extra" units="extra" default="0.9" multipleValues="1" nearestValue="0" >0.0,0.1,0.2,0.3,0.4,0.5,0.6,0.7,0.8,0.9</Dimension>
```
becomes now:
```xml
<Dimension name="extra" units="extra" default="0.9" multipleValues="1" nearestValue="0" >0,0.1,0.2,0.3,0.4,0.5,0.6,0.7,0.8,0.9</Dimension>
```

Reply:
I think not all errors are the result of this branch. When I run the functional tests on master using postgresql, I get similar errors. Running against sqlite (default) works fine. I've tracked it down to a difference when selecting a `real` from psql vs sqlite.

We currently store all dim values that are a `CDF_FLOAT` or `CDF_DOUBLE` as a `real` in the database.

When a dimension value like `10.0` is stored as real in sqlite, and you `select` it, you get `10.0`.
If you store `10.0` in postgresql, the `select` rounds it to `10`.

We could think about storing these dimension fields as a `numeric (10, 1)`, however I don't know if this fixes all issues or introduces new ones.

This happens when you `select` a real in postgresql:
```
adaguc=# select path, height from t20240420t142720604_h78goam7gcvl33cthua6 where height=10.0;
                                     path                                      | height 
-------------------------------------------------------------------------------+--------
 /adaguc/adaguc-server-master/data/datasets/test_WINS50_windshear_processor.nc |     10
(1 row)

adaguc=# update t20240420t142720604_h78goam7gcvl33cthua6 set height=10.0 where height=10;
UPDATE 1
adaguc=# select path, height from t20240420t142720604_h78goam7gcvl33cthua6 where height=10.0;
                                     path                                      | height 
-------------------------------------------------------------------------------+--------
 /adaguc/adaguc-server-master/data/datasets/test_WINS50_windshear_processor.nc |     10
(1 row)
```
And this is what happens when you select a `real` in sqlite. Second field is the height dimension field, created from the same netcdf. No rounding is done:
```
sqlite> select * from t20240421t140242178_ohzu24f9uf7q1y9k07i1;                                                                                                                                                                                                                           
/adaguc/adaguc-server-master/data/datasets/test_WINS50_windshear_processor.nc|10.0|0|2024-01-18T23:05:50Z|0|-1250.0|-1250.0|541250.0|583750.0|0|1|2|3
/adaguc/adaguc-server-master/data/datasets/test_WINS50_windshear_processor.nc|100.0|1|2024-01-18T23:05:50Z|0|-1250.0|-1250.0|541250.0|583750.0|0|1|2|3
/adaguc/adaguc-server-master/data/datasets/test_WINS50_windshear_processor.nc|200.0|2|2024-01-18T23:05:50Z|0|-1250.0|-1250.0|541250.0|583750.0|0|1|2|3
/adaguc/adaguc-server-master/data/datasets/test_WINS50_windshear_processor.nc|300.0|3|2024-01-18T23:05:50Z|0|-1250.0|-1250.0|541250.0|583750.0|0|1|2|3
/adaguc/adaguc-server-master/data/datasets/test_WINS50_windshear_processor.nc|600.0|4|2024-01-18T23:05:50Z|0|-1250.0|-1250.0|541250.0|583750.0|0|1|2|3

```

Reply:
Resolving this comment:

- The psql vs sqlite issue is now solved by applying an sqlite "formatting hack" to make sqlite and psql consistent.
- I've added a `starttests_psql.sh` which performs the xml replace and runs all tests against psql. It is possible to run it inside Docker as well, if you run adaguc-db through docker compose. I'll describe this part in the documentation.